pkg/service/endpoints_controller.go

```diff
@@ -26,135 +27,280 @@ import (
+
+	// TODO: keep a map of pods to services to handle this condition.
+}
+
+func (e *EndpointController) addPodServicesToQueue(obj interface{}) {
+	pod := obj.(*api.Pod)
+	services, err := e.serviceStore.GetPodServices(*pod)
+	if err != nil {
+		// don't log this error because this function makes pointless
+		// errors when no services match.
+		return
+	}
+	for i := range services {
+		svc := &services[i]
+		glog.Infof("pod %v/%v matches service %v/%v", pod.Namespace, pod.Name, svc.Namespace, svc.Name)
+		e.enqueueService(svc)
 	}
 }
 
-// SyncServiceEndpoints syncs endpoints for services with selectors.
-func (e *EndpointController) SyncServiceEndpoints() error {
-	services, err := e.client.Services(api.NamespaceAll).List(labels.Everything())
+// obj could be an *api.Service, or a DeletionFinalStateUnknown marker item.
+func (e *EndpointController) enqueueService(obj interface{}) {
+	key, err := framework.DeletionHandlingMetaNamespaceKeyFunc(obj)
 	if err != nil {
-		glog.Errorf("Failed to list services: %v", err)
-		return err
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
 	}
-	var resultErr error
-	for i := range services.Items {
-		service := &services.Items[i]
 
-		if service.Spec.Selector == nil {
-			// services without a selector receive no endpoints from this controller;
-			// these services will receive the endpoints that are created out-of-band via the REST API.
-			continue
+	e.queue.Dirty(key)
+}
+
+func (e *EndpointController) worker() {
+	for {
+		key, quit := e.queue.BeginProcessing()
+		if quit {
+			return
 		}
+		e.syncService(key.(string))
+		e.queue.FinishProcessing(key)
+	}
+}
 
-		glog.V(5).Infof("About to update endpoints for service %s/%s", service.Namespace, service.Name)
-		pods, err := e.client.Pods(service.Namespace).List(labels.Set(service.Spec.Selector).AsSelector())
+func (e *EndpointController) syncService(key string) {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing service %q endpoints. (%v)", key, time.Now().Sub(startTime))
+	}()
+	obj, exists, err := e.serviceStore.Store.GetByKey(key)
+	if err != nil || !exists {
+		// Delete the corresponding endpoint, as the service has been deleted.
+		// TODO: Please note that this will delete an endpoint when a
+		// service is deleted. However, if we're down at the time when
+		// the service is deleted, we will miss that deletion, so this
+		// doesn't completely solve the problem. See #6877.
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			glog.Errorf("Need to delete endpoint with key %q, but couldn't understand the key: %v", key, err)
+			return
+		}
+		err = e.client.Endpoints(namespace).Delete(name)
 		if err != nil {
-			glog.Errorf("Error syncing service: %s/%s, skipping", service.Namespace, service.Name)
-			resultErr = err
-			continue
+			glog.Errorf("Error deleting endpoint %q: %v", key, err)
```

Comment:
Shouldn't we requeue this service in case of error? It could fail because it was rate limited.